sstables/sstables.cc

```diff
@@ -2035,11 +2022,13 @@ future<> sstable::snapshot(const sstring& dir) const {
 
 future<> sstable::change_state(sstable_state to, delayed_commit_changes* delay_commit) {
     co_await _storage->change_state(*this, to, _generation, delay_commit);
+    _state = to;
 }
 
 future<> sstable::pick_up_from_upload(sstable_state to, generation_type new_generation) {
     co_await _storage->change_state(*this, to, new_generation, nullptr);
     _generation = std::move(new_generation);
+    _state = to;
 }
```

Comment:
perhaps add a do_change_state(state, gen, delay_commit = nullptr), so we don't have two places changing and updating state.

Reply:
Check the #13018 . The current plan is to extend the existing sstables uploading API so that it could pick up an sstable from local disk and upload it into S3 and it's going to be much more than just state change :)